src/lib/operators/set_union.cpp

```diff
@@ -0,0 +1,371 @@
+  ReferenceMatrix reference_matrix_right;
+
+  build_reference_matrix(_input_table_left(), reference_matrix_left);
+  build_reference_matrix(_input_table_right(), reference_matrix_right);
+
+  /**
+   * Init the virtual pos lists
+   */
+  VirtualPosList virtual_pos_list_left(_input_table_left()->row_count(), 0u);
+  std::iota(virtual_pos_list_left.begin(), virtual_pos_list_left.end(), 0u);
+  VirtualPosList virtual_pos_list_right(_input_table_right()->row_count(), 0u);
+  std::iota(virtual_pos_list_right.begin(), virtual_pos_list_right.end(), 0u);
+
+  /**
+   * Sort the virtual pos lists, so we can merge them
+   */
+  std::sort(virtual_pos_list_left.begin(), virtual_pos_list_left.end(),
+            VirtualPosListCmpContext{reference_matrix_left});
+  std::sort(virtual_pos_list_right.begin(), virtual_pos_list_right.end(),
+            VirtualPosListCmpContext{reference_matrix_right});
+
+  /**
+   * Build result table
+   */
+  auto left_idx = size_t{0};
+  auto right_idx = size_t{0};
+  const auto num_left = virtual_pos_list_left.size();
+  const auto num_right = virtual_pos_list_right.size();
+
+  // Somewhat random way to decide on a chunk size.
+  const auto out_chunk_size = std::max(_input_table_left()->chunk_size(), _input_table_right()->chunk_size());
+
+  auto out_table = Table::create_with_layout_from(_input_table_left(), out_chunk_size);
+
+  std::vector<std::shared_ptr<PosList>> pos_lists(reference_matrix_left.size());
+  std::generate(pos_lists.begin(), pos_lists.end(), [&] { return std::make_shared<PosList>(); });
+
+  // Adds one row to the pos_lists currently being filled
+  const auto emit_row = [&](const ReferenceMatrix& reference_matrix, size_t row_idx) {
+    for (size_t pos_list_idx = 0; pos_list_idx < pos_lists.size(); ++pos_list_idx) {
+      pos_lists[pos_list_idx]->emplace_back(reference_matrix[pos_list_idx][row_idx]);
+    }
+  };
+
+  // Turn 'pos_lists' into a new chunk and append it to the table
+  const auto emit_chunk = [&]() {
+    Chunk chunk;
+
+    for (size_t pos_lists_idx = 0; pos_lists_idx < pos_lists.size(); ++pos_lists_idx) {
+      const auto segment_column_id_begin = _column_segment_begins[pos_lists_idx];
+      const auto segment_column_id_end = pos_lists_idx >= _column_segment_begins.size() - 1
+                                             ? _input_table_left()->column_count()
+                                             : _column_segment_begins[pos_lists_idx + 1];
+      for (auto column_id = segment_column_id_begin; column_id < segment_column_id_end; ++column_id) {
+        auto ref_column = std::make_shared<ReferenceColumn>(
+            _referenced_tables[pos_lists_idx], _referenced_column_ids[column_id], pos_lists[pos_lists_idx]);
+        chunk.add_column(ref_column);
+      }
+    }
+
+    out_table->emplace_chunk(std::move(chunk));
+  };
+
+  // Comparator of two ReferenceMatrix rows
+  const auto cmp = [](const auto& matrix_a, const auto idx_a, const auto& matrix_b, const auto idx_b) {
```

Comment:
If you gave the lambda a more speaking name and had `ReferenceMatrix` as a type, the comment would be unnecessary

Reply:
more speaking name it is